### PR TITLE
improve PR reviews

### PR DESCRIPTION
**Pull request by akosyakov**
#### What it does

I want to start working on improving our review process. We have many PRs and I feel being a bottleneck in the process when many things can be checked without me. This PR proposes:
- the PR template with a checklist which should be filled in before a PR can be merged. It does not matter by whom (contributor or reviewers). 
- more strict rules on approves to make sure that testing and design validation gets higher priority than human linting of empty lines. It should also make it easier for reviewers to judge whether some aspects already reviewed and improve knowledge sharing by nudging developers to test and review design.

[Rendered PR guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md)

#### How to test

Review whether something missing, too strict or unnecessary in the proposed review checklist

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully reviewed following [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/ak/pr_template/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -1 +1,32 @@
...
+  - use `git pull -r` to pick up changes from the master
+- [ ] tests/benchmarks are defined for functionallity/perf optimizations and verified by a reviewer, not an author
+- [ ] [changelog](https://github.com/theia-ide/theia/blob/master/CHANGELOG.md) is updated
+- [ ] breaking changes are justified and recorded in the changelog
+- [ ] each new file has a proper copyright with the current year and an author as an entity
+- [ ] new dependendencies are justified and [verified](https://github.com/theia-ide/theia/wiki/Registering-CQs#wip---new-ecd-theia-intellectual-property-clearance-approach-experimental)
+- [ ] copied code are justified and [approved via a CQ](https://github.com/theia-ide/theia/wiki/Registering-CQs#case-3rd-party-project-code-copiedforked-from-another-project-into-eclipse-theia-maintained-by-us)
+- [ ] new code is aligned with [project organization](https://github.com/theia-ide/theia/wiki/Code-Organization) and [coding conventions](https://github.com/theia-ide/theia/wiki/Coding-Guidelines)
+- [ ] each review approve is provided with a comment regarding whether functionallity is tested, design or code are aligned with project organization and conventions
```
Comment (tsmaeder) [inline .github/PULL_REQUEST_TEMPLATE.md]: The last point in the check list ("approve comment") seem to imply that the reviewer built and tested the code. If this is required, I would make it an explicit item in the check list. One could still review the code, but not approve.

Would it make sense to require instructions on how to test the changed behaviour? I figure it's not always obvious how to do that: we had some such cases implementing VS Code API.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -1 +1,32 @@
...
+<!-- Describe how a reviewer can verify changes within Theia repo -->
+
+#### Review Checklist
+<!-- For completed items, change [ ] to [x].  -->
+<!-- For inapplicable items, cross out with `[ ] foo` to `[x] ~~foo~~`.  -->
+
+- [ ] commits are signed-off: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md#sign-your-work
+- [ ] commit history is cleaned up of merge and fix up commits
+  - use `git pull -r` to pick up changes from the master
+- [ ] tests/benchmarks are defined for functionallity/perf optimizations and verified by a reviewer, not an author
```
Comment (tsmaeder) [inline .github/PULL_REQUEST_TEMPLATE.md]: Does this mean we have to have unit tests? If yes, what coverage? What about things that are easier to test manually (typos, adding a UI element like menu item)?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -1 +1,32 @@
...
+#### How to test
+<!-- Describe how a reviewer can verify changes within Theia repo -->
+
+#### Review Checklist
+<!-- For completed items, change [ ] to [x].  -->
+<!-- For inapplicable items, cross out with `[ ] foo` to `[x] ~~foo~~`.  -->
+
+- [ ] commits are signed-off: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md#sign-your-work
+- [ ] commit history is cleaned up of merge and fix up commits
+  - use `git pull -r` to pick up changes from the master
```
Comment (tsmaeder) [inline .github/PULL_REQUEST_TEMPLATE.md]: So in other words, rebase your PR on the latest state of master, right? I do that graphically in Eclipse, t.b.h.

**Comment by tsmaeder** [comment]
I know it's a lot of work for you doing all those reviews, but I think it has served us well, so far. You have made sure the architectural integrity has been preserved. I fear the "design validation" could start to suffer. It's always hard to tell someone to start over when they've invested 2 weeks on the implementation. I think doing threads like this one https://spectrum.chat/theia/dev/how-should-we-go-about-register-monaco-commands-used-by-vs-code-extensions~2c74f56c-2d24-4f21-9db6-8041f0e861c9 will help with validating a design before we get too invested in them.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -1 +1,32 @@
...
+#### What it does
+<!-- Include relevant issues and describe how they are addressed -->
+
+#### How to test
+<!-- Describe how a reviewer can verify changes within Theia repo -->
+
+#### Review Checklist
+<!-- For completed items, change [ ] to [x].  -->
+<!-- For inapplicable items, cross out with `[ ] foo` to `[x] ~~foo~~`.  -->
+
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply .github/PULL_REQUEST_TEMPLATE.md]: I would put this first. It might seem obvious but I feel that having to check this explicitly might motivate to go through it once more or extra carefully.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -1 +1,32 @@
...
+- [ ] commits are signed-off: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md#sign-your-work
+- [ ] commit history is cleaned up of merge and fix up commits
```
Comment (svenefftinge) [inline .github/PULL_REQUEST_TEMPLATE.md]: commit history is rebased on master and contains only meaningful commits and changes (less are usually better)

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (svenefftinge) [reply .github/PULL_REQUEST_TEMPLATE.md]: Can we leave this out?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -1 +1,32 @@
...
+  - use `git pull -r` to pick up changes from the master
+- [ ] tests/benchmarks are defined for functionallity/perf optimizations and verified by a reviewer, not an author
+- [ ] [changelog](https://github.com/theia-ide/theia/blob/master/CHANGELOG.md) is updated
+- [ ] breaking changes are justified and recorded in the changelog
+- [ ] each new file has a proper copyright with the current year and an author as an entity
```
Comment (svenefftinge) [inline .github/PULL_REQUEST_TEMPLATE.md]: each new file has a proper copyright with the current year and the name of contributing entity (individual or company)

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -1 +1,32 @@
...
+- [ ] new dependendencies are justified and [verified](https://github.com/theia-ide/theia/wiki/Registering-CQs#wip---new-ecd-theia-intellectual-property-clearance-approach-experimental)
+- [ ] copied code are justified and [approved via a CQ](https://github.com/theia-ide/theia/wiki/Registering-CQs#case-3rd-party-project-code-copiedforked-from-another-project-into-eclipse-theia-maintained-by-us)
```
Comment (svenefftinge) [inline .github/PULL_REQUEST_TEMPLATE.md]: copied code *is`

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (akosyakov) [reply .github/PULL_REQUEST_TEMPLATE.md]: I don't want to make the process overly strict and did not mean require unit tests for each PR.

I meant that each PR should at least explain how to reproduce a bug, test new functionality, verify perf improvements. If someone wants to add unit or UI tests it is always welcomed, but not mandatory.

I will remove it and add `A contributor should explain how to reproduce a bug, test new functionality or verify perf improvements.` as a comment under `How to test` section.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 1 above)
Comment (akosyakov) [reply .github/PULL_REQUEST_TEMPLATE.md]: > The last point in the check list ("approve comment") seem to imply that the reviewer built and tested the code. If this is required, I would make it an explicit item in the check list. One could still review the code, but not approve.

sounds good 

> Would it make sense to require instructions on how to test the changed behaviour? I figure it's not always obvious how to do that: we had some such cases implementing VS Code API.

A contributor should think about it and fill in `How to test` section, no? How he can know otherwise that he is doing the right thing.

Comment (akosyakov) [reply .github/PULL_REQUEST_TEMPLATE.md]: for VS Code API one should create a vscode extension and test it against VS Code and Theia, or write a unit test against VS Code APIs. Do you mean that we should have such recommendations?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -1 +1,30 @@
...
+<!-- Include relevant issues and describe how they are addressed. -->
+
+#### How to test
+<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify perf improvements.  -->
+
+#### Review Checklist
+<!-- For completed items, change [ ] to [x].  -->
+<!-- For inapplicable items, cross out with `[ ] foo` to `[x] ~~foo~~`.  -->
+
+- [ ] as an author I have carefully reviewed and thoroughly tested my changes.
```
Comment (kittaakos) [inline .github/PULL_REQUEST_TEMPLATE.md]: as an author, I have

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -1 +1,30 @@
...
+- [ ] commits are signed-off: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md#sign-your-work
+- [ ] commit history is rebased on master and contains only meaningful commits and changes (less are usually better)
+  - for example use `git pull -r` or `git fetch && git rebase` to pick up changes from the master
+- [ ] [changelog](https://github.com/theia-ide/theia/blob/master/CHANGELOG.md) is updated
+- [ ] breaking changes are justified and recorded in the changelog
+- [ ] each new file has a proper copyright with the current year and the name of contributing entity (individual or company)
```
Comment (kittaakos) [inline .github/PULL_REQUEST_TEMPLATE.md]: each new file has proper

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -1 +1,30 @@
...
+- [ ] new dependendencies are justified and [verified](https://github.com/theia-ide/theia/wiki/Registering-CQs#wip---new-ecd-theia-intellectual-property-clearance-approach-experimental)
```
Comment (kittaakos) [inline .github/PULL_REQUEST_TEMPLATE.md]: dependencies

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 11]
```diff
@@ -1 +1,30 @@
...
+- [ ] copied code is justified and [approved via a CQ](https://github.com/theia-ide/theia/wiki/Registering-CQs#case-3rd-party-project-code-copiedforked-from-another-project-into-eclipse-theia-maintained-by-us)
+- [ ] new code is aligned with [project organization](https://github.com/theia-ide/theia/wiki/Code-Organization) and [coding conventions](https://github.com/theia-ide/theia/wiki/Coding-Guidelines)
```
Comment (kittaakos) [inline .github/PULL_REQUEST_TEMPLATE.md]: the new code is aligned

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 12]
```diff
@@ -1 +1,30 @@
...
+- [ ] each review approve is provided with a comment regarding whether functionallity is tested, design or code are aligned with project organization and conventions
```
Comment (kittaakos) [inline .github/PULL_REQUEST_TEMPLATE.md]: the functionality (note the typo not just the missing `the`)

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 13]
```diff
@@ -1 +1,30 @@
...
+  - an approve without comment should be dismissed
```
Comment (kittaakos) [inline .github/PULL_REQUEST_TEMPLATE.md]: the approve without a comment

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 14]
```diff
@@ -1 +1,30 @@
...
+Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
+-->
+
+#### What it does
+<!-- Include relevant issues and describe how they are addressed. -->
+
+#### How to test
+<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify perf improvements.  -->
+
+#### Review Checklist
```
Comment (kittaakos) [inline .github/PULL_REQUEST_TEMPLATE.md]: I would go with `Review checklist` as we do not use a capital for the other titles either.

[minor]: perhaps you could start each bullet point with a capital, and finish the sentence with a period. But it is up to you.

**Comment by akosyakov** [comment]
Please have a look again. I've tried to address all comments and did following changes:
- beginning of the review checklist clarifies responsibilities of an author and reviewers
- following points are ordered according to the importance, starting from testing and design checking and finishing with proper git history
- inapplicable items should be just marked as completed as well, crossing them out is big effort with the GH UI

**Comment by akosyakov** [comment]
> I know it's a lot of work for you doing all those reviews, but I think it has served us well, so far. You have made sure the architectural integrity has been preserved. I fear the "design validation" could start to suffer. It's always hard to tell someone to start over when they've invested 2 weeks on the implementation. I think doing threads like this one https://spectrum.chat/theia/dev/how-should-we-go-about-register-monaco-commands-used-by-vs-code-extensions~2c74f56c-2d24-4f21-9db6-8041f0e861c9 will help with validating a design before we get too invested in them.

Agree that contributors should start discussions and open PRs early even if they are incomplete in order to verify the design first.

**Comment by benoitf** [comment]
it would be great to use https://www.conventionalcommits.org as well

**Comment by akosyakov** [comment]
@benoitf let's discuss it next dev meeting?

The goal of this PR to make existing review expectations clear.

Also @benoitf had some good points here https://github.com/theia-ide/theia/pull/5622#discussion_r300280864. I try to summarise here:
- the PR should be minimal as possible or at least make unnecessary clean ups as separate commits
- description should be complete
- such review checklist can be long

Regarding the last point. It looks already long to me as well. I still think we should have it to make it clear about expectations in order to help new contributors to know what to look for. Maybe we better to add `Reviewing Pull Request` guide as a separate document and here have a short checklist like:
#### Review checklist
- [ ] as an author, I have thoroughly tested my changes and carefully reviewed in accordance with [the review guideline](a link to guide here).
- [ ] each approve has supporting comments in accordance with [the review guideline](a link to guide here).
  - the approve without a comment should be dismissed

**Comment by benoitf** [comment]
@akosyakov sure I don't want to hijack the thread, let's discuss it on dev meeting

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 15]
```diff
@@ -1 +1,31 @@
...
+
+#### How to test
+<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify perf improvements.  -->
+
+#### Review checklist
+<!-- For completed items, change [ ] to [x].  -->
+<!-- For inapplicable items, mark them as completed as well.  -->
+
+- [ ] as an author, I have carefully reviewed and thoroughly tested my changes
+- [ ] each approve has supporting comments regarding some points below
```
[the remark this answers is not in the document]
Comment (svenefftinge) [reply .github/PULL_REQUEST_TEMPLATE.md]: Can we remove them and have a general section on how to do code reviews? Or is there a template for code reviews as well?

[the remark this answers is not in the document]
Comment (akosyakov) [reply .github/PULL_REQUEST_TEMPLATE.md]: not implemented yet: https://github.community/t5/How-to-use-Git-and-GitHub/Pull-Request-Review-Template/td-p/6857/page/2

I will split.

**Comment by akosyakov** [comment]
Ideas to cover in the review guideline:
- collective code ownership
  - anyone can change anything
  - anyone should review to learn and ensure that code is readable and align with project expectations
  - does not mean that requirements, design and conventions should be disregarded
  - one who understands rules and design that he can maintain it, becomes a de-facto maintainer (code owner)
- ...

**Comment by akosyakov** [comment]
> Can we remove them and have a general section on how to do code reviews? Or is there a template for code reviews as well?

I've extracted a new document on how to collaborate and review PRs. The PR template is short now and references this document.

File: doc/pull-requests.md [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+- [ ] as an author, I have thoroughly tested my changes and carefully reviewed following [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
+
+#### Reminder for reviewers
+
+- as a reviewer, I agree to review in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
+
+```
+
+<a name="design-review"></a>
+- [2.](#design-review) A PR can be opened early for the design review before going into the detail implementation.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [2.](#design-review) A PR can be opened early for the design review before going into the detailed implementation.
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 17]
```diff
@@ -1 +1,21 @@
...
+the requirements below.
+
+Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
+-->
+
+#### What it does
+<!-- Include relevant issues and describe how they are addressed. -->
+
+#### How to test
+<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify perf improvements.  -->
```
Comment (thegecko) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements.  -->
```

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 18]
```diff
@@ -1 +1,21 @@
...
+
+#### Review checklist
+
+- [ ] as an author, I have thoroughly tested my changes and carefully reviewed following [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
```
Comment (thegecko) [inline .github/PULL_REQUEST_TEMPLATE.md]:
```suggestion
- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
```

File: doc/pull-requests.md [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+the requirements below.
+
+Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
+-->
+
+#### What it does
+<!-- Include relevant issues and describe how they are addressed. -->
+
+#### How to test
+<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify perf improvements.  -->
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements.  -->
```

File: doc/pull-requests.md [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+
+#### Review checklist
+
+- [ ] as an author, I have thoroughly tested my changes and carefully reviewed following [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [ ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
```

File: doc/pull-requests.md [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+<a name="review-reqs"></a>
+- [1.](#review-reqs) A review can be requested when:
+  - [The PR template](#pr-template) is filled in.
+  - Changes are thoroughly tested by an author.
+  - Changes thoroughly reviewed following the [review checklist](#review-checklist) by an author.
+<a name="review-request-gh"></a>
+- [2.](#review-request-gh) A review can be requested explicitly using GitHub: https://help.github.com/en/articles/requesting-a-pull-request-reviewed
+<a name="review-request-comment"></a>
+- [3.](#review-request-comment) A review can be also requested as a comment from any GitHub users.
+  - For example to invite a person originally filed an issue for testing.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
  - For example to invite the person who originally filed an issue for testing.
```

File: doc/pull-requests.md [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+
+## Review Checklist
+
+<a name="checklist-build-and-test"></a>
+- [1.](#checklist-build-and-test) The new code is built and tested according to `How to test` section of a PR description.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [1.](#checklist-build-and-test) The new code is built and tested according to the `How to test` section of a PR description.
```

File: doc/pull-requests.md [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+<a name="checklist-project-org"></a>
+- [2.](#checklist-project-org) The new code is aligned with [project organization](https://github.com/theia-ide/theia/wiki/Code-Organization) and [coding conventions](https://github.com/theia-ide/theia/wiki/Coding-Guidelines).
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [2.](#checklist-project-org) The new code is aligned with the [project organization](https://github.com/theia-ide/theia/wiki/Code-Organization) and [coding conventions](https://github.com/theia-ide/theia/wiki/Coding-Guidelines).
```

File: doc/pull-requests.md [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+<a name="checklist-commit-history"></a>
+- [9.](#checklist-commit-history) Commit history is rebased on master and contains only meaningful commits and changes (less are usually better).
+  - For example, use `git pull -r` or `git fetch && git rebase` to pick up changes from the master.
+
+## Reviewing
+
+<a name="reviewing-template"></a>
+- [1.](#eviewing-template) Reviewers should check that a PR has a [proper description](#pr-template).
+<a name="reviewing-fn"></a>
+- [2.](#eviewing-fn) Reviewers should build and verify changes according to `How to test` section of a PR description.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [2.](#eviewing-fn) Reviewers should build and verify changes according to the `How to test` section of a PR description.
```

File: doc/pull-requests.md [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+<a name="reviewing-checklist"></a>
+- [3.](#reviewing-checklist) Reviewers should ensure that all checks from [the review checklist](#review-checklist) are successful.
+<a name="reviewing-share"></a>
+- [4.](#reviewing-share) A reviewer does not need to ensure everything but can verify a part of it and provide feedback as a comment.
+
+### Requesting Changes
+
+<a name="changes-review-reqs"></a>
+- [1.](#changes-review-reqs) Changes should be requested if an author does not follow [review requirements](#review-reqs).
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [1.](#changes-review-reqs) Changes should be requested if an author does not follow the [review requirements](#review-reqs).
```

File: doc/pull-requests.md [hunk 26]
```diff
@@ -0,0 +1,155 @@
...
+<a name="changes-no-nit"></a>
+- [2.](#changes-no-nit) Changes cannot be requested because of the personal preferences of a reviewer.
+  - Such change requests should be dismissed.
+<a name="changes-no-out-of-scope"></a>
+- [3.](#changes-no-out-of-scope) Changes cannot be requested if they address issues out of the scope of a PR.
+  - Such change requests should be dismissed and an issue should be filed to address them separately.
+<a name="changes-style-agreement"></a>
+- [4.](#changes-style-agreement) Styles and coding preferences should not be discussed on the PR, but raised to [the dev meeting](https://github.com/theia-ide/theia/wiki/Dev-Meetings),
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [4.](#changes-style-agreement) Styles and coding preferences should not be discussed on the PR, but raised in [the dev meeting](https://github.com/theia-ide/theia/wiki/Dev-Meetings),
```

File: doc/pull-requests.md [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+  agreed by the team, applied to [the coding guidelines](https://github.com/theia-ide/theia/wiki/Coding-Guidelines) and after that followed by all contributors.
+
+### Approving
+
+<a name="justifiying-approve"></a>
+- [1.](#justifiying-approve) Each approve should have supporting comments following these guidelines.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [1.](#justifiying-approve) Each approval should have supporting comments following these guidelines.
```

File: doc/pull-requests.md [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+<a name="dismissing-approve"></a>
+- [2.](#dismissing-approve) An approve without a comment should be dismissed.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
- [2.](#dismissing-approve) An approval without a comment should be dismissed.
```

File: doc/pull-requests.md [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+<a name="suggesting-help-on-pr"></a>
+- [3.](#suggesting-help-on-pr) Reviewers have to suggest his help via a comment to avoid intervening in an author work.
+<a name="landing-stale-pr"></a>
+- [4.](#landing-stale-pr) Such comment is not required if an author is not responsive.
+
+## Landing
+
+<a name="landing-pr"></a>
+- [1.](#landing-pr) A PR can be landed when:
+  - CI build is succeeded.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
  - CI build has succeeded.
```

File: doc/pull-requests.md [hunk 30]
```diff
@@ -0,0 +1,155 @@
...
+  - An author accepted the [Eclipse Contributor Agreement](https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md#eclipse-contributor-agreement).
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
  - The author has accepted the [Eclipse Contributor Agreement](https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md#eclipse-contributor-agreement).
```

File: doc/pull-requests.md [hunk 31]
```diff
@@ -0,0 +1,155 @@
...
+  - All checks from [the review checklist](#pull-request-review-checklist) are approved at least by one reviewer.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
  - All checks from [the review checklist](#pull-request-review-checklist) are approved by at least one reviewer.
```

File: doc/pull-requests.md [hunk 32]
```diff
@@ -0,0 +1,155 @@
...
+  - There are no unresolved review comments.
+
+## Reverting
+
+<a name="reverting-pr"></a>
+- [1.](#reverting-pr) If a PR causes regressions after landing
+then an author has 2 days to resolve them after that a PR has to be reverted.
```
Comment (thegecko) [inline doc/pull-requests.md]: > has 2 days

How will this be managed?

File: doc/pull-requests.md [hunk 33]
```diff
@@ -0,0 +1,155 @@
...
+
+## Closing
+
+<a name="closing-pr"></a>
+- [1.](#closing-pr) A reviewer cannot close a PR without a reason.
+<a name="closing-pr-reasons"></a>
+- [2.](#closing-pr-reasons) A PR may be closed, for example, because of the following reasons:
+  - It introduces a functionality which should be implemented as external Theia or VS Code extensions.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
  - It introduces functionality which should be implemented as external Theia or VS Code extensions.
```

File: doc/pull-requests.md [hunk 34]
```diff
@@ -0,0 +1,155 @@
...
+  - It introduces structural and API changes between core extensions.
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
  - It introduces structural or API changes between core extensions.
```

File: doc/pull-requests.md [hunk 35]
```diff
@@ -0,0 +1,155 @@
...
+  Such changes have to be done by an experienced maintainer to avoid regressions and long reviews.
+  - It should be a 3rd party component, e.g. Theia is not a logging framework or a proxy server.
+  - It changes development infrastructure, e.g. testing frameworks, packaging and so on.
+  Such changes have to be done by active maintainers after agreement on [the dev meeting](https://github.com/theia-ide/theia/wiki/Dev-Meetings).
```
Comment (thegecko) [inline doc/pull-requests.md]:
```suggestion
  Such changes have to be done by active maintainers after agreement in [the dev meeting](https://github.com/theia-ide/theia/wiki/Dev-Meetings).
```

File: doc/pull-requests.md (hunk 32 above)
Comment (akosyakov) [reply doc/pull-requests.md]: We will ask an author to look into a fix if there is no response and we don't have time to fix it ourself then we revert it. As it was done here: https://github.com/theia-ide/theia/pull/5770

Comment (akosyakov) [reply doc/pull-requests.md]: Also if an author is responsive and trying to fix, but we keep receiving bug reports from users as it was the case with SCM PR, then we will revert everything and wait for the better PR.

**Comment by akosyakov** [comment]
@thegecko thank for looking at it, I've applied your suggestions

File: doc/pull-requests.md (hunk 32 above)
Comment (akosyakov) [reply doc/pull-requests.md]: Do you think we need to rephrase it somehow to make it more clear?

**Comment by akosyakov** [comment]
@marcdumais-work @svenefftinge please have a look whether you happy with the current state

File: doc/pull-requests.md [hunk 36]
```diff
@@ -0,0 +1,155 @@
...
+<a name="checklist-sign-off"></a>
+- [8.](#checklist-sign-off) Commits are signed-off: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md#sign-your-work.
+<a name="checklist-commit-history"></a>
+- [9.](#checklist-commit-history) Commit history is rebased on master and contains only meaningful commits and changes (less are usually better).
+  - For example, use `git pull -r` or `git fetch && git rebase` to pick up changes from the master.
+
+## Reviewing
+
+<a name="reviewing-template"></a>
+- [1.](#eviewing-template) Reviewers should check that a PR has a [proper description](#pr-template).
```
Comment (marcdumais-work) [inline doc/pull-requests.md]: Suggestion: also make it clear that the commit(s) should have a proper description (at minimum equivalent to `What it does` from the PR template. Other info from filled-in template could be nice-to-have as well). This is missed sometimes and makes it harder to understand what a commit does without digging-in the code.

Comment (marcdumais-work) [reply doc/pull-requests.md]: Maybe this would fit better in the "review checklist" section, around \#8 / \#9

Comment (akosyakov) [reply doc/pull-requests.md]: @marcdumais-work i think it should go to review checklist, could you suggest wording.

Also @benoitf suggested to use conventional commits. I think it is a good idea in the sense that we should automate as much as possible from this checklist. With conventional commits we will need only ask to provide proper commit messages and then we can generate changelogs from them. I know that electon follow such approach. It could be worth to look into after landing this PR.

Comment (marcdumais-work) [reply doc/pull-requests.md]: >  could you suggest wording.

- [10.]() Each commit has meaningful title and a body that explains what it does. One can take inspiration from the "what it does" section from the PR.

> Also @benoitf suggested to use conventional commits. [...] It could be worth to look into after landing this PR.

:+1:

**Comment by akosyakov** [comment]
pinged Sven offline, he said that he is fine to follow it, merging
